ouster-ros/src/os_sensor_node.cpp

```diff
@@ -0,0 +1,695 @@
+            throw;
+        }
+
+        RCLCPP_INFO_STREAM(get_logger(),
+                           "Sensor " << hostname << " configured successfully");
+    }
+
+    bool load_config_file(const std::string& config_file,
+                          sensor::sensor_config& out_config) {
+        std::ifstream ifs{};
+        ifs.open(config_file);
+        if (ifs.fail()) return false;
+        std::stringstream buf;
+        buf << ifs.rdbuf();
+        out_config = sensor::parse_config(buf.str());
+        return true;
+    }
+
+    // fill in values that could not be parsed from metadata
+    void populate_metadata_defaults(sensor::sensor_info& info,
+                                    sensor::lidar_mode specified_lidar_mode) {
+        if (!info.name.size()) info.name = "UNKNOWN";
+
+        if (!info.sn.size()) info.sn = "UNKNOWN";
+
+        ouster::util::version v = ouster::util::version_of_string(info.fw_rev);
+        if (v == ouster::util::invalid_version)
+            RCLCPP_WARN(
+                get_logger(),
+                "Unknown sensor firmware version; output may not be reliable");
+        else if (v < sensor::min_version)
+            RCLCPP_WARN(
+                get_logger(),
+                "Firmware < %s not supported; output may not be reliable",
+                to_string(sensor::min_version).c_str());
+
+        if (!info.mode) {
+            RCLCPP_WARN(
+                get_logger(),
+                "Lidar mode not found in metadata; output may not be reliable");
+            info.mode = specified_lidar_mode;
+        }
+
+        if (!info.prod_line.size()) info.prod_line = "UNKNOWN";
+
+        if (info.beam_azimuth_angles.empty() ||
+            info.beam_altitude_angles.empty()) {
+            RCLCPP_ERROR(
+                get_logger(),
+                "Beam angles not found in metadata; using design values");
+            info.beam_azimuth_angles = sensor::gen1_azimuth_angles;
+            info.beam_altitude_angles = sensor::gen1_altitude_angles;
+        }
+    }
+
+    // try to write metadata file
+    bool write_metadata(const std::string& meta_file,
+                        const std::string& metadata) {
+        std::ofstream ofs(meta_file);
+        if (ofs.is_open()) {
+            ofs << metadata << std::endl;
+            ofs.close();
+            RCLCPP_INFO_STREAM(get_logger(), "Wrote sensor metadata to " <<
+                meta_file);
+        } else {
+            RCLCPP_WARN(
+                get_logger(),
+                "Failed to write metadata to %s; check that the path is valid. "
+                "If you provided a relative path, please note that the working "
+                "directory of all ROS nodes is set by default to $ROS_HOME",
+                meta_file.c_str());
+            return false;
+        }
+        return true;
+    }
+
+    void create_publishers() {
+        rclcpp::SensorDataQoS qos;
```

Comment:
is qos configurability a future feature or is it standard to just use SensorDataQoS

Reply:
I saw this PR by chance and wanted to add a quick comment. 
For some use cases it would actually be nice to adjust the QoS to a more reliable one such as ParameterEventsQoS() via a parameter in the params.yaml.

Specifically, we record lidar_packets on a Raspberry Pi without Pointcloud2 for performance reasons. Later, on a more powerful machine, we convert lidar_packets to Pointcloud2 by replaying the bagfile and starting the Ouster driver again. For this conversion we usually change the QoS to ParameterEventsQoS() to get a more reliable result without data drops.

Reply:
Currently, the choice of QoS policy is hard-coded but the plan is to give users the ability to select and apply the policy of their choice.